Tweak `self` arg not as first argument of a method diagnostic

estebank opened the pull request:
Mention that `self` is only valid on "associated functions"
```
error: unexpected `self` argument in function
  --> $DIR/self-in-function-arg.rs:1:15
   |
LL | fn foo(x:i32, self: i32) -> i32 { self }
   |               ^^^^ not valid as function argument
   |
   = note: `self` is only valid as the first argument of an associated function
```

When it is a method, mention it must be first
```
error: unexpected `self` argument in function
  --> $DIR/trait-fn.rs:4:20
   |
LL |     fn c(foo: u32, self) {}
   |                    ^^^^ must be the first associated function argument
```

Move a bunch of error recovery methods to `diagnostics.rs` away from `parser.rs`.

Fix #51547. CC #60015.

rust-highfive commented:
r? @pnkfelix

(rust_highfive has picked a reviewer for you, use r? to override)

estebank commented:
r? @Centril parser code shuffle

petrochenkov commented on src/libsyntax/parse/diagnostics.rs:
```diff
@@ -60,6 +146,449 @@ impl RecoverQPath for Expr {
 }
 
 impl<'a> Parser<'a> {
+    pub fn look_ahead<R, F>(&self, dist: usize, f: F) -> R where
```
This function is not diagnostics specific.

Centril commented on src/libsyntax/parse/diagnostics.rs:
```diff
@@ -501,6 +1030,141 @@ impl<'a> Parser<'a> {
+    ) -> PResult<'a, ast::Arg> {
+        let sp = arg.pat.span;
+        arg.ty.node = TyKind::Err;
+        let mut err = self.struct_span_err(sp, "unexpected `self` argument in function");
```
```suggestion
        let mut err = self.struct_span_err(sp, "unexpected `self` parameter in function");
```

Centril commented on src/libsyntax/parse/diagnostics.rs:
```diff
@@ -501,6 +1030,141 @@ impl<'a> Parser<'a> {
+        arg.ty.node = TyKind::Err;
+        let mut err = self.struct_span_err(sp, "unexpected `self` argument in function");
+        if is_trait_item {
+            err.span_label(sp, "must be the first associated function argument");
```
```suggestion
            err.span_label(sp, "must be the first associated function parameter");
```

Centril commented on src/libsyntax/parse/diagnostics.rs:
```diff
@@ -501,6 +1030,141 @@ impl<'a> Parser<'a> {
+        if is_trait_item {
+            err.span_label(sp, "must be the first associated function argument");
+        } else {
+            err.span_label(sp, "not valid as function argument");
```
```suggestion
            err.span_label(sp, "not valid as function parameter");
```

Centril commented on src/libsyntax/parse/diagnostics.rs:
```diff
@@ -501,6 +1030,141 @@ impl<'a> Parser<'a> {
+            err.span_label(sp, "must be the first associated function argument");
+        } else {
+            err.span_label(sp, "not valid as function argument");
+            err.note("`self` is only valid as the first argument of an associated function");
```
```suggestion
            err.note("`self` is only valid as the first parameter of an associated function");
```

petrochenkov commented on src/libsyntax/parse/diagnostics.rs:
```diff
@@ -60,6 +146,449 @@ impl RecoverQPath for Expr {
+        if self.expected_tokens.is_empty() {
+            if self.token == *t {
+                self.bump();
+                Ok(false)
```
Functions like this have a small "good" path (which doesn't belong to diagnostics) and the remaining big part belonging to diagnostics.
In this case perhaps it's reasonable to split them into two functions, one in `parse.rs` and another in `diagnostics.rs`.

Centril commented on src/libsyntax/parse/diagnostics.rs:
```diff
@@ -60,6 +146,449 @@ impl RecoverQPath for Expr {
+                     .map_or(String::new(), |t| t.to_string());
+            i.enumerate().fold(b, |mut b, (i, a)| {
+                if tokens.len() > 2 && i == tokens.len() - 2 {
+                    b.push_str(", or ");
```
Refactor out `b.push_str(connective)`?

Centril commented on src/libsyntax/parse/diagnostics.rs:
```diff
@@ -60,6 +146,449 @@ impl RecoverQPath for Expr {
+    }
+
+    pub fn fatal(&self, m: &str) -> DiagnosticBuilder<'a> {
+        self.sess.span_diagnostic.struct_span_fatal(self.span, m)
```
```suggestion
        self.span_fatal(self.span, m)
```

Centril commented:
I'll do a more thorough review later.

@estebank Can you split the "move stuff" and "change diagnostics output" into separate commits?

estebank commented:
@Centril addressed all review comments but one, which I think should be part of a separate PR cleaning up all the times this has been done in the parser.

bors commented:
:umbrella: The latest upstream changes (presumably #61105) made this pull request unmergeable. Please resolve the merge conflicts.

bors commented:
:umbrella: The latest upstream changes (presumably #61150) made this pull request unmergeable. Please resolve the merge conflicts.

estebank commented on src/libsyntax/parse/diagnostics.rs:
```diff
@@ -542,6 +1017,138 @@ impl<'a> Parser<'a> {
+        is_trait_item: bool,
+    ) -> PResult<'a, ast::Arg> {
+        let sp = arg.pat.span;
+        arg.ty.node = TyKind::Err;
```
Because we don't _know_ wether this is any of the following:

- a function with an arbitrary argument called `self`
- a method with `self` not at the start
- a method with `self` at the start _and_ a second `self`

I thought it'd be safer to just mark it as `TyKind::Err` and avoid potential incorrect diagnostics elsewhere.

Centril replied:
Good insight; on the other hand, I'd say that `self` not at the start is probabilistically more correct than the others... but, in any case, let's land this as-is. :)

estebank commented:
@bors r=Centril

bors commented:
:pushpin: Commit 4e68ddca90a8f3ee580445b1fc83a6a6d0c7f94f has been approved by `Centril`

<!-- @bors r=Centril 4e68ddca90a8f3ee580445b1fc83a6a6d0c7f94f -->
